### ISSUE TITLE
Valhalla 3.1 is not thread safe

### ISSUE THREAD

**Issue by dnesbitt61** [valhalla/valhalla#2867]
After updating our embedded Valhalla application to Valhalla 3.1, we started seeing intermittent crashes. This happened during multi-threaded route calls. As many threading issues are, this issue was intermittent (hard to replicate) and demonstrated different crash symptoms. After a thorough code-review I found PR #2772 which offers a fix. The problem is, this refers to an undocumented build option and the default configuration builds Valhalla in a non thread safe manner. So my conclusion is that Valhalla 3.1 is not thread-safe by default. Fixes include:

1. This option (if kept) must be documented in the build instructions.
2. The default configuration MUST be with thread safe reference counters enabled. Valhalla is an open source project and its use should be safe with default build configurations. A note on when it is safe to set this to OFF can be attached for those who are sure they do not require thread safety. 
3. Is this build option even necessary? What is the penalty for always using a thread safe reference counter? It seems to me that unless there is a real performance penalty that the code should always be configured to be thread safe.

**Comment by dnesbitt61** [comment]
It seems that changing from unsafe reference counter (basically an int) to a thread-safe reference counter leads to a performance drop of perhaps 5% in the simple benchmark I ran. So this is worth preserving for applications where one can guarantee single threaded behavior will always apply.

Is there a way to inject this at run-time in cases where global_synchronized_cache is set to true? Is that the only time it would be needed?

**Comment by kevinkreiser** [comment]
ha you beat me to it but i was in the middle of typing:

I'm trying to reason about this but im not 100% convinced I know what is going on. Here's some information.

So we used to have a raw pointer to graphtiles exposed via graphreader through the configured tilecache. Raw pointers were of course not threadsafe but also the pointer itself was not responsible for its lifetime. We then made a change to valhalla that allowed the returned tiles to manage their own lifetime. We could have done so with std::shared_ptr but this comes with an additional 5% performance penalty (because of locking). The problem here is we cant get the whole RAII design to work without a tradeoff. Basically if you want to have multiple threads use the globally synchronized tilecache then you need to have thread safe smart pointers because they will race over destructing the tiles. But if you are running in any other mode paying 5% performance for locking that is useless doesnt make sense either. Seems like it was a mistake to have tiles own their own lifetimes.

I think the only way we could inject this at runtime would be with inheritance. Where the global_synchronized_cache returns a child class of the regular graph tile that is synchronized.

**Comment by dnesbitt61** [comment]
I am fine with keeping a build option but it should be reversed. By default we should build with thread safety. Only those who know they are building a single threaded application would they set the option to use a non-thread-safe reference counter. At any rate, this needs to be documented.

**Comment by kevinkreiser** [comment]
looking at it a little closer i think it shouldnt be that hard... i'll try to make a PR tonight

**Comment by kevinkreiser** [comment]
yeah it seems like we can implement the two policies directly inside of `GraphTile` and not have it inherit from `intrusive_ref_counter`. we'd need a flag inside of graphtile that switches between incrementing and decrementing an atomic vs non atomic counter and we'd need to set this flag in the constructor of GraphTile (with a parameter). then when we call `GraphTile::Create*` from the `global_synchronized_cache` we can hopefully pass this parameter through and boost's intrusive pointer will call our own increment/decrement functions inside of GraphTile which wil have a simple if on that flag to do one increment vs the other.

maybe there is some pitfall here but i think the general idea that im describing here would work to do this at runtime

**Comment by xlqian** [comment]
We recently bumped Valhalla, on which our application is dependent, to 3.1.0. I quickly spotted the crash by activating the sanitizer. Thank you for telling how to fix it.
